doc/fast-reboot/Fast-reboot_Flow_Improvements_HLD.md

```diff
@@ -0,0 +1,158 @@
+This new flag can be used later on for any functionality which we want to start only after init flow finished in case of fast-reboot.
+This is to prevent interference in the fast-reboot reconciliation process and impair the performance, for example enablement of flex counters.
+
+References:
+https://github.com/Azure/SONiC/blob/master/doc/fast-reboot/fastreboot.pdf
+https://github.com/Azure/sonic-swss/blob/master/neighsyncd/restore_neighbors.py
+https://github.com/Azure/sonic-buildimage/blob/master/dockers/docker-orchagent/enable_counters.py
+
+# 2 Functional Requirements
+
+The new Fast-reboot design should meet the following requirments:
+
+- Reboot the switch into a new SONiC software version using kexec.
+- Upgrade the FW by the new SONiC image if needed.
+- Recover all applications state with the new image to the previous state prior the reboot.
+- Recover ASIC state after reset to the previous state prior the reboot.
+- Recover the Kernel state after reset to the previous state prior the reboot.
+- Sync the Kernel and ASIC with changes on the network which happen during fast-reboot.
+- Control plane downtime will not exceed 90 seconds.
+- Data plane downtime will not exceed 30 seconds.
+
+# 3 Use Cases
+
+## 3.1 In-Service restart and upgrade
+
+### SWSS docker
+
+When swss docker starts with the new kernel, all the port/LAG, vlan, interface, arp and route data should be restored from CONFIG DB, APP DB, Linux Kernel and other reliable sources. There could be ARP, FDB changes during the restart window, proper sync processing should be performed.
+
+### Syncd docker
+
+The restart of syncd docker should leave data plane intact until it starts again with the new kernel. After restart, syncd configures the HW with the state prior the reboot by all network applications.
+
+# 4 Reconciliation at syncd
+
+## 4.1 Orchagent Point Of View
+
+When orchagent starts with the new SONiC image, the same infrastructure we use to reconcile fastfast-boot will start.
+After INIT_VIEW and create_switch functions sent to syncd (reset of the ASIC took place here), 'warmRestoreAndSyncUp' will be executed.
+This function will populate m_toSync with all tasks for syncd, by APP DB and CONFIG DB prior the reboot.
+To verify orchagent reached the same state as before the reboot, 'warmRestoreValidation' will verify no pending tasks left in the queue, meaning all proccessed succesfully and in the pipeline for syncd to configure the HW.
+At the end APPLY_VIEW will be sent to syncd to finalize the process, from this point orchagent enter the main loop and operates normally.
+
+### NOTICE
+
+'warmRestoreValidation' might fail the operation just like in fastfast-reboot case, if the way orchagent process an event from the DB is handled differently with the new software version the task will fail to execute and fast-reboot will fail along with it.
+This is solvable by the db migrator.
```

Comment:
First this is based on fastfast-reboot infra and not warm-reboot infra, so temp view is not used at all and INIT/APPLY view does nothing basically.
For DB migration and OID mismatches, this could be resolved by a modification to the way we backup the DB.
This should also address the problem when fast rebooting from a EOS to SONiC, as we discussed on mail.
From the mail thread: 

__If I understand correctly, we basically want to keep the json dump files we use on the current design instead of a Redis rdb backup file and it will be backward compatible along with this scenario as well.
So I am thinking maybe instead of taking a backup copy of Redis DB, keep the current approach we push these dumps to the DB with swssconfig tool by swssconfig.sh.
We can still use the fastfast-reboot reconciliation logic on each network application and orchagent since they will start after, and all data will be present in the DB at this point.
What do you think?_